### ISSUE TITLE
add a figurine notation

### ISSUE THREAD

**Issue by braoult** [mliebelt/PgnViewerJS#248]
I made a simple test, using UTF-8 chess for pieces symbols instead of language based letters (KQRBN) :
https://github.com/braoult/PgnViewerJS/commit/9b6f49a3eebaa47ec357795d07823c16fefce1ad
https://github.com/braoult/PgnViewerJS/commit/084ba7f87c0d811e4d5daae88a8594dc9d7ffe54

Obviously, if we want to display symbols, we should not go this way, and instead use maybe a toggle (e.g. visualName: true/false). Or not, see Lichess section below.

The result: it works, but difficult to read (I use Noto Sans):
![image](https://user-images.githubusercontent.com/8958429/121243546-1ce3a100-c89e-11eb-9de7-5637fa32657f.png)

Meantime, I had a look at Lichess display system, they use two fonts:

- https://lichess1.org/assets/_KLITAX/font/lichess.woff2
- https://lichess1.org/assets/_KLITAX/font/lichess.chess.woff2

I converted them to ttf, to understand what they are:

The first one is for some strange symbols, not interesting.
The second one is very interesting, it is a "Noto Sans" family font, with only a few characters defined, all those used in chess moves notation (with english names KQRBN:

`1-9, a-h, K, Q, R, B, N, hyphen, endash (the hyphen for castle move I suppose), =, +, @, O (for castle move)`

![image](https://user-images.githubusercontent.com/8958429/121237453-5c5abf00-c897-11eb-816b-00ccba4b0b5b.png)

So they simply replace the KQRBN by symbols, interesting approach, no need anymore to take care languages.
The drawback of doing that is that they cannot display figures differently (with letters).

I think we could have a similar approach, or maybe use the reserved chess figures UTF-8 characters within the font we choose (I like Noto Sans, I use it on my desktop :-)

The figures they did use are pretty simple, easy to view:
Lichess bishop:
![image](https://user-images.githubusercontent.com/8958429/121241024-3c2cff00-c89b-11eb-8d0b-59c66f7cae0f.png)

Default Noto Sans bishop:
![image](https://user-images.githubusercontent.com/8958429/121243623-32f16180-c89e-11eb-956c-23bb35b5ca5e.png)

I cannot continue today, but it is may be something we could consider for PgnViewer ?

**Comment by mliebelt** [comment]
Thank you for your input. This is a duplicate of #134 

Your ticket has way more input, that could be used later for an implementation. I see some more work todo, and would like to structure it a little bit:

* Find the right font. The font of Lichess is ok, the one you have shown at the end is no option in my opinion.
* How about integrating it together with internationalization (which should continue to work)?
* Is it necessary to have different symbols for white and black (which breaks the current approach of Lichess)?

I would like to check different chess sites, to see if we can see a different approach. And to collect them and compare then.

**Comment by braoult** [comment]
> Your ticket has way more input, that could be used later for an implementation. I see some more work todo, and would like to structure it a little bit:

>     * Find the right font. The font of Lichess is ok, the one you have shown at the end is no option in my opinion.

Exact. In fact it is the same "Noto Sans" as Lichess, if I understand my Firefox Dev Tool, except they did replace some characters.

>     * How about integrating it together with internationalization (which should continue to work)?

Yes, I was thinking of keeping it along the current "Language-based" letters, and adding a separate set for symbol display, with a switch (true/false). Some languages could have this set by default, if they don't have a simple abbreviation character to design pieces (Japanese, Arabic, Hebrew, ...).

>     * Is it necessary to have different symbols for white and black (which breaks the current approach of Lichess)?

No, we need only one. My 2 commits show the changes with the 2 sets, (white / black) that are available in UTF-8.
So a possible solution could be to replace one of these sets with something similar to Lichess's own figurines.

My knowledge on web font did not allow me (yet) to use Lichess font, along Noto Sans with PngViewer. It looks like they use "Noto Sans", and override some part of it. My Browser says the font in use is "Noto Chess" and "Noto Sans" for Lichess, I would like to reproduce this at least for a test.

I don't know at all how it works.

Also, Lichess' source code is huge, and I did not find (yet) the source code for their fonts (and associated licenses).

What I know is that Google's Noto fonts are free (as french "Libre"), with a [SIL Open Font License (OFL)](https://scripts.sil.org/cms/scripts/page.php?site_id=nrsi&id=OFL).

**Comment by mliebelt** [comment]
I did a first functional trial, and it was much easier than I thought. The footprint is now bigger (+ 20k), but there are 2 chess fonts bundled with it: lichess (ugly), and alpha (much better). The changes to Javascript were minimal, so I commit and push them, and create a bundle, so that others are able to download it from https://mliebelt.github.io/PgnViewerJS/lib/pgnv.js

If it would be possible to create a (minimal) font with only the figures needed, mapped to B, K, N, P, R, Q, it would be awesome. Or people could do their own figurine notation, just by adding 2 CSS rules.

![image](https://user-images.githubusercontent.com/520039/121392583-d189cb00-c94f-11eb-9a1d-a97468767dd6.png)

**Comment by braoult** [comment]
I was just able to make the test (my git repo was completely broken, and spent ages trying to fix it :-).

Your changes work perfectly, thanks !

Here are screenshots from [my site](https://www.raoult.com/chess), with the 3 possible figurine display so far: lichess/alpha/native UTF8 :

Lichess font:
![lichess](https://user-images.githubusercontent.com/8958429/121500589-bca84880-c9de-11eb-91dc-2f1183e7b1e3.png)

Alpha font:
![alpha](https://user-images.githubusercontent.com/8958429/121500649-cc279180-c9de-11eb-8551-65c00495f6dc.png)

Native Noto Sans font:
![default-noto-sans](https://user-images.githubusercontent.com/8958429/121500728-de093480-c9de-11eb-9f5f-96c76f2e393e.png)

However, I am unable to build pgn-viewer anymore from a fresh install from your repo. Are there any steps relative to the last version different that `npm install` in `root`, and `npm build` in `modules/{pgn-reader,pgn-viewer}` ? If it looks abnormal, I will open a ticket with the error log.

**Comment by mliebelt** [comment]
Thank you a lot for your feedback! I will try to do a fresh checkout, and then trigger the 3 builds as usual. I am not aware of any change. This will take some time, due to other duties I have, but hopefully this evening, I will be able to do a short test.

**Comment by braoult** [comment]
> So this is what I do all the time, with a fresh install (just repeated to be sure):
>     * Go into the root directory.
>     * Do a `npm install`.
>     * Change to `modules/pgn-viewer`.
>     * Do a `npm install`.

Oh. In fact, what I did was what is described in installation doc :
* Go into the root directory
* Do a `npm install`: OK
* Go in `modules/pgn-reader`
* Do a `npm run-script build` : OK
* Go in modules/pgn-viewer
* Do a `npm run-script build`: NOK (8 errors)

**Comment by mliebelt** [comment]
Ok, I tried now different versions of node, combined with different versions of npm. All of them work for me: V10.23, v14.17, v15.8

All of them work for me. Downloaded a zip of the current state, and used all  the time both variants: `npm run build`, and `npm run-script build`.

And as said, to build pgn-viewer, you don't have to build pgn-reader. Please document the version you are using (node, npm).

**Comment by mliebelt** [comment]
When trying to reproduce your problem, I found that the current build does not work with the current version of Chessground (7.12.0). See #253 for some context. So I had to fix the Chessground version to 7.11.1 (exactly that one, not greater).

Could you please download again a zip file, and see if that fixes your problem? Sorry, I was not aware of a breaking change of Chessground. My local version 7.11.1 worked, but switching to 7.12.0 (and if you do a fresh download, you will get the newest version), this problem arises.

**Comment by braoult** [comment]
> All of them work for me. Downloaded a zip of the current state, and used all the time both variants: `npm run build`, and `npm run-script build`.
> And as said, to build pgn-viewer, you don't have to build pgn-reader. Please document the version you are using (node, npm).

My versions of node and npm:
```
$ node --version
v12.18.2
$ npm --version
6.14.8
```

About chessground, I can confirm the error messages were related to it.
It is working now (with latest clone) with the following :
```
$ git clone https://github.com/mliebelt/PgnViewerJS.git
$ cd PgnViewerJS
$ npm install
$ cd modules/pgn-viewer
$ npm install
$ npm run-script build
```

Sorry, I don't know anything about node/npm, so I can just follow blindly what I am told to do, I cannot really help when a problem occurs, except maybe sending error log ;-)
